### PR TITLE
(fix)(utils/uri) change decode behavior for handle json/url values

### PR DESCRIPTION
**Pull request by steffgrez**
Fix problem when client send json/full url percent-encoded in a query string.
Now make decode in parse_query_string rather than in the requests processing.

ex:
url=http%3A%2F%2Ffalconframework.org%3Ftest%3D1
json=%7B%22test1%22%3A%20%22data1%22%2C%20%22test2%22%3A%20%22data2%22%7D

### CODE REVIEW

File: falcon/request.py [hunk 1]
```diff
@@ -220,10 +220,9 @@ def __init__(self, env, options=None):
 
         # PERF(kgriffs): if...in is faster than using env.get(...)
         if 'QUERY_STRING' in env:
-            query_str = env['QUERY_STRING']
+            self.query_string = env['QUERY_STRING']
```
Comment (kgriffs) [inline falcon/request.py]: With this change L230-L231 are no longer necessary. Could you remove those lines?

File: falcon/util/uri.py [hunk 2]
```diff
@@ -292,9 +292,9 @@ def parse_query_string(query_string, keep_blank_qs_values=False):
...
             old_value = params[k]
             if isinstance(old_value, list):
-                old_value.append(v)
+                old_value.append(decode(v))
             else:
-                params[k] = [old_value, v]
+                params[k] = [old_value, decode(v)]
 
         else:
             if ',' in v:
```
Comment (kgriffs) [inline falcon/util/uri.py]: Seems like v must be decoded before checking if it contains commas, since the commas might be encoded. What do you think about factoring out the `decode(v)` and placing it above L290, e.g.:

```python
v = decode(v)

for k in params:
    # ...
```

File: tests/test_query_params.py [hunk 3]
```diff
@@ -49,7 +49,7 @@ def test_simple(self):
         self.assertEqual(store['limit'], '25')
 
     def test_percent_encoded(self):
-        query_string = 'id=23%2c42&q=%e8%b1%86+%e7%93%a3'
+        query_string = 'id=23,42&q=%e8%b1%86+%e7%93%a3'
```
Comment (kgriffs) [inline tests/test_query_params.py]: Could you please revert this change? See also my other note above re L300.

File: falcon/util/uri.py (hunk 2 above)
Comment (steffgrez) [inline falcon/util/uri.py]: In fact, making this factoring will don't fix original problem.

The only thing to determine if a comma is the reserved character or a simple character in your value is the percent-encoding. So, if i want to make a comma-separated list, i use the reserved character ','. Otherwise, if i want to send a value with a comma (like json, complex url, etc ...), i use  percent-encoding.
For me, falcon has to decode value at the last moment. So query parser will not handle all percent-encoded comma. Like that it won't confound with this sub-delimiter.

ex:
I want send the json '{"test1": "data1", "test2": "data2"}' by query parameter.
So we encode it : %7B%22test1%22%3A%20%22data1%22%
if we decode value before checking commas, the result will be a list: ['{"test1": "data1"', ' "test2": "data2"}'] while we wait a string.

As i understand URI rfc:
http://tools.ietf.org/html/rfc3986#section-2.4
"When a URI is dereferenced, the components and subcomponents significant to the scheme-specific dereferencing process (if any) must be parsed and separated before the percent-encoded octets within those components can be safely decoded, as otherwise the data may be mistaken for component delimiters"
http://tools.ietf.org/html/rfc3986#section-7.3
"Applications must split the URI into its components and subcomponents prior to decoding the octets, as otherwise the decoded octets might be mistaken for delimiters."

File: falcon/request.py (hunk 1 above)
Comment (steffgrez) [inline falcon/request.py]: it makes sense, Request class don't use self.query_string elsewhere without make a test.

File: tests/test_query_params.py (hunk 3 above)
Comment (steffgrez) [inline tests/test_query_params.py]: See my response above.

File: falcon/util/uri.py (hunk 2 above)
Comment (kgriffs) [inline falcon/util/uri.py]: Ah, good point. Makes sense. Thanks for the clarification! Would you mind adding a `# NOTE(steffgrez): blah blah` comment to the code about this?

Comment (steffgrez) [inline falcon/util/uri.py]: ok for me

File: falcon/util/uri.py [hunk 4]
```diff
@@ -311,7 +314,7 @@ def parse_query_string(query_string, keep_blank_qs_values=False):
                     # will be the same for 'foo=1,,3' as 'foo=1&foo=&foo=3'.
                     v = [element for element in v if element]
```
Comment (kgriffs) [inline falcon/util/uri.py]: Technically, there's a breaking change here, although there isn't currently a test written for it. If individual elements are percent-encoded, they will no longer be decoded here. Do you think it would be worth doing the following? If so, it would be good to also add a test. I suspect few (if any) apps require this, but it may be better to be safe than sorry.

```python
if not keep_blank_qs_values:
    v = [decode(element) for element in v if element]
else:
    v = [decode(element) for element in v]
```

The alternative would be to add a "breaking change" notice to the commit message (and eventually the changelog when we release 0.3).

File: falcon/util/uri.py (hunk 2 above)
Comment (rogerbush8) [inline falcon/util/uri.py]: >> Seems like v must be decoded before checking if it contains commas, since the
>> commas might be encoded. What do you think about factoring out the decode(v)
>> and placing it above L290, e.g.:

This is incorrect.  The purpose of the % encoding is to "turn off the special nature" of the character.  It's just like backslashing a character in a C string.  So if we decode prior to splitting, we are in trouble.

The procedure should be:
1. Split on literal "&" to split the n=v parameter pairs.
2. For each of these parameter pairs, subsplit on ",", which is simply the next delimiter.
3. Now decode each fragment separately.  This means:
    a) decode the name of a parameter pair (e.g. a utf-8 funny char in an identifier...god help us)
    b) decode the value:
        i) Single value
       ii) Multiple values (split on ","), each of which must be decoded.

3.b.ii. also shows how someone needs to do encoding if they have multi-valued parameters separated by "," (they must encode each part separately, and NOT encode the ",").  Note that we are EXTENDING the meaning of "," to have a special meaning in the same spirit that the original spec had of "&".  "," special meaning is - turn this value into an array, and use "," as the subdelimiter.
   

v = decode(v)

if k in params:
    # ...

File: falcon/util/uri.py (hunk 4 above)
Comment (rogerbush8) [inline falcon/util/uri.py]: Individual elements in an array induced by "," should be decoded IMHO.  It's just recursive design logic.  Why would one type of value be encoded/decoded, whereas another would not?  This is counter-intuitive and makes no sense.  Values are encoded so that we can pass the funny characters without having issues.  Just go to each element in the array and decode it.

I know it sounds un-fast.  But first make it correct, then make it fast.

Comment (steffgrez) [inline falcon/util/uri.py]: In fact, i didn't see this case, and for me, it is not a "breaking change" but an error. I think too that handle this case is the thing to do. So i fix it and i add also unit tests for later.

another question, do we accept a mix like that : a=1&a=2&a=3,4 ?
it's a little weird, but the case could exist ... Actually, falcon doesn't handle this case.

File: falcon/util/uri.py (hunk 2 above)
Comment (kgriffs) [inline falcon/util/uri.py]: At this point, I think we're just missing the decoding of the name (`k`), and improving the documentation per (3.b.ii).

@steffgrez Would you mind adding a line to decode `k` and a test for the same? Regarding (3.b.ii), we could update the note in the docstring: For example:

```
     Note:
         In addition to the standard HTML form-based method for specifying
         lists by repeating a given param multiple times, Falcon supports
         a more compact format in which the param may be given a single time
         but set to a ``list`` of comma-separated elements (e.g., 'foo=a,b,c').

         When using this format, if any values need to be uri-encoded, the 
         client should take care not to encode the commas used as 
         delimiters. Otherwise, Falcon will not treat the value as a list.
 
         The two different ways of specifying lists may not be mixed in
         a single query string for the same parameter.
```

**Comment by EvanLyu** [comment]
In my case,
Request content-type is `application/x-www-form-urlencoded`
Raw data is "RESULT_CODE=000000&EXTEND=CARD_TYPE%3D20%26CARD_ID%3DMp8eLZ&CHECK_CODE=FEEFAF749"

If i use falcon version 0.2 ,
falcon parser result is:
{
    RESULT_CODE: 000000,
    EXTEND: CARD_TYPE=20,
    CARD_ID: Mp8eLZ,
    CHECK_CODE: FEEFAF749
}

But real parsed data is:
{
    RESULT_CODE: 000000,
    EXTEND: CARD_TYPE=20&CARD_ID=Mp8eLZ,
    CHECK_CODE: FEEFAF749
}

It deal my problem.

**Comment by kgriffs** [comment]
@steffgrez Once my last two docstring comments have been addressed, I think this PR is ready for final review. Would you mind squashing all patches down to a single commit, in preparation for merging? Also, note that the commit message's first line should begin with `fix`, rather than `(fix)`. Please also mention the following at the bottom of the commit description on a line by itself:

`Fixes #369`

Thanks!

**Comment by steffgrez** [comment]
done.  

Thanks for your feedback !

**Comment by kgriffs** [comment]
Nice work! LGTM
